File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;
        int closest_sum = Integer.MAX_VALUE / 2; // To avoid overflow
        int current_sum = 0;

        for (int i = 0; i < n; i++) {
            int left = i + 1;
            int right = n - 1;

            while (left < right) {

                current_sum = nums[i] + nums[left] + nums[right];

                if (Math.abs(current_sum - target) < Math.abs(closest_sum - target)) {
                    closest_sum = current_sum; // compare & assign small value to closest
                }

                if (current_sum < target)
                    left++;
                else if (current_sum > target)
                    right--;
                else {
                    return current_sum; // equals to target
                }
            }
        }
        return closest_sum;

    }
}

File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        size = len(nums)
        closer = 200000
        nums.sort()

        for i in range(size - 1):

          if i > 0 and nums[i - 1] == nums[i]:
            continue
          
          left = i + 1
          right = size - 1

          while left < right:

            total = nums[i] + nums[left] + nums[right]
            abs_current_diff = abs(total - target)
            abs_closer_diff = abs(closer - target)

            if abs_current_diff < abs_closer_diff:
              closer = total

            if total > target:
              right -= 1
            elif total < target:
              left += 1
            else:
              return total
          
        return closer

            
            